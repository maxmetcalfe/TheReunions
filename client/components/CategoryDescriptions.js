import React, { Component } from "react";

class CategoryDescriptions extends Component {

  constructor() {
    super();
    this.baseClass = "coin";
    this.descriptions = {
      five: {
        value: 5,
        text: "All four members present + special occasion"
      },
      four: {
        value: 4,
        text: "All four members present"
      },
      three: {
        value: 3,
        text: "Three members present"
      }, 
      two: {
        value: 2,
        text: "Two members present"
      }
    };
  }

  render() {
    return (
      <table className="category-descriptions">
        <tbody>
          {Object.keys(this.descriptions).map((category) => (
            <tr key={category}>
              <td><div className={this.baseClass + " " + category}>{this.descriptions[category].value}</div></td>
              <td><h3 className={category}>{this.descriptions[category].text}</h3></td>
            </tr>
          ))}
        </tbody>
      </table>
    );
  }
}

export default CategoryDescriptions;
