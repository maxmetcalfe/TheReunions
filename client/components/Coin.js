import React, { Component } from "react";
import PropTypes from "prop-types";
import resetMap from "../utils/resetMap"
import constants from "../constants";

var CLASS_NAME = "coin";
var SELECTED_CLASS_NAME = "coin selected";

class Coin extends Component {

  constructor(props) {
    super(props);
    this.reunions = props.reunions;
  }

  componentDidMount() {
    this.map = document.getElementById("map");
  }

  selectMemberReunionsByType(e) {
    // Unselect all reunions on the map.
    resetMap(this.map);
    // Send the selected element to the parent App.
    this.props.setSelection(this);
    // Tag this coin as selected.
    this.setState({selected: true});
    var self = this;
    e.reunions.forEach(function(reunion) {
      if (constants.CATEGORIES[reunion.properties.category] === self.props.type) {
        var marker = document.getElementById(reunion.properties.element_id);
        marker.classList.add("picked");
      }
    })
  }

  className() {
    if (this.displayAsSelected()) {
      return SELECTED_CLASS_NAME + " " + this.props.type;
    }
    return CLASS_NAME + " " + this.props.type;
  }

  displayAsSelected() {
    if (this.props.selection && this.props.id === this.props.selection.props.id) {
      return true;
    }
    return false;
  }

  render() {
    return (
      <div className={this.className()} onClick={()=>{this.selectMemberReunionsByType(this.props)}}>{this.props.counts}</div>
    );
  }
}

Coin.propTypes = {
  counts: PropTypes.number,
  reunions: PropTypes.array,
  selection: PropTypes.object,
  setSelection: PropTypes.func,
  type: PropTypes.string.isRequired,
  id: PropTypes.string
};

export default Coin;
