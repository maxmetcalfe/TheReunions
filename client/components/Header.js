import React, { Component } from "react";

class Header extends Component {

  render() {
      return (
        <div>
          <h1>Welcome to The Reunions.</h1>
          <h2>
            A group of four friends travel together around the world.
            This is where they have been.
          </h2>
        </div>
      );
  }
}

export default Header;
