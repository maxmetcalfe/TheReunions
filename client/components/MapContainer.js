import React, { Component } from "react";
import PropTypes from "prop-types";
import MemberTable from './MemberTable';
import constants from "../constants";
import resetMap from "../utils/resetMap"

const STARTING_CENTER = new mapkit.Coordinate(37.697948, -97.314835)

class MapContainer extends Component {

  constructor() {
    super();
    this.state = { panelVisible: true };
    window.mapContainer = this;
  }

  componentDidMount() {
    mapkit.init({
      authorizationCallback: function(done) {
        fetch("/api/token")
        .then(res => res.text())
        .then(done);
      },
    });

    var map = new mapkit.Map("map", {
      center: STARTING_CENTER,
      colorScheme: mapkit.Map.ColorSchemes.Dark
    });

    this.map = map;
    this.mapElement = document.getElementById("map");

    var annotations = [];
    this.props.reunions.features.forEach(function(reunion) {
      if (!reunion.properties.location) {
        return;
      }
      var annotationCoordinate = new mapkit.Coordinate(parseFloat(reunion.geometry.coordinates[1]), parseFloat(reunion.geometry.coordinates[0]));
      var annotation = new mapkit.MarkerAnnotation(annotationCoordinate, {
        title: reunion.properties.location,
        color: constants.MARKER_COLORS[reunion.properties.category],
        glyphColor: constants.MARKER_GLYPH_COLOR,
        data: {
          category: constants.CATEGORIES[reunion.properties.category],
          members: reunion.properties.members
        }
      })
      annotations.push(annotation);
    })
    map.showItems(annotations);
  }

  selectReunions(selection) {
    var selectedAnnotations;
    if (selection.category) {
      selectedAnnotations = this.map.annotations.filter(function(annotation) { return annotation.data.category === selection.category });
    } else if (selection.member) {
      selectedAnnotations = this.map.annotations.filter(function(annotation) { return annotation.data.members.indexOf(selection.member.toUpperCase())});
    }
    this.map.showItems(selectedAnnotations);
  }

  hidePanel() {
    this.setState({ panelVisible: false });
    this.mapElement.classList.add("full");
    this.map.resize();
  }

  showPanel() {
    this.setState({ panelVisible: true });
    this.mapElement.classList.remove("full");
    this.map.resize();
  }

  render() {
      if (this.props.selection && this.mapElement) {
          resetMap(this.mapElement);
          this.selectReunions(this.props.selection);
      }
      return (
        <div id="map-container">
          <div id="map"></div>
          <div id="show-panel-button" onClick={this.showPanel.bind(this)}>SHOW PANEL</div>
          <div id="panel" className={this.state.panelVisible ? "side-panel" : "side-panel hide"}>
            <MemberTable summaryCounts={this.props.summaryCounts}
              reunionsForMember={this.props.reunionsForMember}
              selection={this.props.selection}
            />
            <div id="panel-slider" onClick={this.hidePanel.bind(this)}>
              <div></div>
            </div>
           </div>
        </div>
      );
  }
}

MapContainer.propTypes = {
  summaryCounts: PropTypes.object.isRequired,
  reunionsForMember: PropTypes.object.isRequired,
  selection: PropTypes.object,
  reunions: PropTypes.object
}

export default MapContainer;
