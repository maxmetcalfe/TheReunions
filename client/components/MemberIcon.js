import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { selectReunions } from "../actions";

var CLASS_NAME = "member-icon";
var SELECTED_CLASS_NAME = "member-icon selected";

export class MemberIcon extends Component {

  constructor(props) {
    super(props);
    this.reunions = props.reunions;
  }

  selectMemberReunions() {
    this.props.dispatch(selectReunions({ member: this.props.member, element: this }));
  }

  displayAsSelected() {
    if (this.props.selection && this.props.id === this.props.selection.element.props.id) {
      return true;
    }
    return false;
  }

  render() {
    var member = this.props.member.toUpperCase();
    return (
      <div className={this.displayAsSelected() ? SELECTED_CLASS_NAME : CLASS_NAME} onClick={()=>{this.selectMemberReunions(this.props)}}>
        {member}
      </div>
    );
  }
}

MemberIcon.propTypes = {
  reunions: PropTypes.array,
  member: PropTypes.string,
  id: PropTypes.string,
  selection: PropTypes.object,
  dispatch: PropTypes.func
}

const mapStateToProps = (dispatch) => {
  return { dispatch: dispatch };
};

export default connect(
  mapStateToProps
)(MemberIcon);

