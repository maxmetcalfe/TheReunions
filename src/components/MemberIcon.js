import React, { Component } from "react";

class MemberIcon extends Component {

  render() {
    return (
      <div className="member-icon">
        {this.props.text}
      </div>
    );
  }
}

export default MemberIcon;
