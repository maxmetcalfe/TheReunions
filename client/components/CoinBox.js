import React, { Component } from "react";
import PropTypes from "prop-types";
import Coin from './Coin';

class CoinBox extends Component {
  
  render() {
    var divKey = this.props.member + "-" + this.props.type;
    return (
      <div key={divKey} className="coin-box">
        {Object.keys(this.props.counts).map((type) => (
            <Coin key={this.props.member + "-" + type}
              type={type}
              id={this.props.member + "-" + type}
              counts={this.props.counts[type]}
              reunions={this.props.reunions}
              selection={this.props.selection}
              setSelection={this.props.setSelection}
            />
        ))}
      </div>
    );
  }
}

CoinBox.propTypes = {
  member: PropTypes.string.isRequired,
  type: PropTypes.string,
  counts: PropTypes.object.isRequired,
  reunions: PropTypes.array.isRequired,
  selection: PropTypes.object,
  setSelection: PropTypes.func
}

export default CoinBox;
