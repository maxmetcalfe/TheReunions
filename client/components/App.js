import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Header from './Header';
import MapContainer from './MapContainer';
import CategoryDescriptions from './CategoryDescriptions';
import ReunionList from './ReunionList';
import { dataLoaded } from "../actions";

const MAP_CONTAINER_BUTTON_CLASS = "map-container-button";
const MAP_CONTAINER_CLOSE_BUTTON_CLASS = "map-container-close-button";

class App extends Component {

  constructor(props) {
    super(props);
    this.state = { mapContainerVisible: false };
    window.app = this;
  }

  displayMapContainer() {
    this.setState({ mapContainerVisible: true });
  }

  hideMapContainer() {
    this.setState({ mapContainerVisible: false });
  }

  mapContainerHiddenLayout(hasData) {
    let reunions = this.props.data && this.props.data.data && this.props.data.data.reunions;
    let showMapButton = hasData ? <div className={MAP_CONTAINER_BUTTON_CLASS} onClick={this.displayMapContainer.bind(this)}>SHOW MAP</div> : null;

    return (
      <div>
        <Header />
        <CategoryDescriptions />
        <ReunionList reunions={reunions}/>
        {showMapButton}
      </div>
    );
  }

  mapContainerVisibleLayout() {
    let reunions = this.props.data && this.props.data.data && this.props.data.data.reunions;
    let reunionsForMember = this.props.data && this.props.data.data && this.props.data.data.reunionsForMember;
    let summaryCounts = this.props.data && this.props.data.data && this.props.data.data.summaryCounts;
    let selection = this.props.data && this.props.data.selection;

    return (
      <div>
        <a href="#" className={MAP_CONTAINER_CLOSE_BUTTON_CLASS} onClick={this.hideMapContainer.bind(this)}></a>
        <MapContainer summaryCounts={summaryCounts}
          reunions={reunions}
          reunionsForMember={reunionsForMember}
          selection={selection}
        />
      </div>
    );
  }

  render() {

    // If we don't have any data yet, fetch it.
    // FIX ME: Why is data double nested?
    let hasData = this.props.data && Object.keys(this.props.data).length > 0;
    if (!hasData) {
      fetch("/api/data")
        .then((response) => response.json())
        .then((data) => this.props.dispatch(dataLoaded(data)))
        .catch((e) => console.warn("Failed to load data: " + e))
    }

    if (this.state.mapContainerVisible) {
      return this.mapContainerVisibleLayout();
    } else {
      return this.mapContainerHiddenLayout(hasData);
    }
  }
}

const mapStateToProps = (data, dispatch) => {
  return { data, dispatch: dispatch };
};

App.propTypes = {
  data: PropTypes. object.isRequired,
  dispatch: PropTypes.func
};

export default connect(
  mapStateToProps
)(App);
