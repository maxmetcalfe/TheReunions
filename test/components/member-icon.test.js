import React from "react"
import renderer from "react-test-renderer"
import { MemberIcon } from "../../client/components/MemberIcon"

const reunions = {
  features: [
    {
      properties: {
        name: "Boston 2014",
        element_id: "Boston2014four",
        location: "Boston",
        category: "four"
      }
    },
    {
      properties: {
        name: "Charleston 2012",
        element_id: "Boston2014three",
        location: "Charleston",
        category: "three"
      }
    },
    {
      properties: {
        name: "New York 2011",
        element_id: "Boston2014two",
        location: "New York",
        category: "two"
      }
    }
  ]
};

const summaryCounts = {
    "j":{"four": 5, "three": 4, "two": 4},
    "m":{"four": 5, "three": 3, "two": 3},
    "d":{"four": 5, "three": 2, "two": 2},
    "g":{"four": 5, "three": 1, "two": 1},
}

const reunionsForMember = {
  "d": [
    { "category": "three", "element_id": "june2018THREE-Boston-1", "location": "Boston ", "name": "June 2018" },
    { "category": "two", "element_id": "july2018TWO-Boston-1", "location": "Boston ", "name": "July 2018" },
    { "category": "one", "element_id": "august2018ONE-Boston-1", "location": "Boston ", "name": "August 2018" }
  ],
  "g": [
    { "category": "three", "element_id": "june2018THREE-Boston-1", "location": "Boston ", "name": "June 2018" },
    { "category": "two", "element_id": "july2018TWO-Boston-1", "location": "Boston ", "name": "July 2018" },
    { "category": "one", "element_id": "august2018ONE-Boston-1", "location": "Boston ", "name": "August 2018" }
  ],
  "j": [
    { "category": "three", "element_id": "june2018THREE-Boston-1", "location": "Boston ", "name": "June 2018" },
    { "category": "two", "element_id": "july2018TWO-Boston-1", "location": "Boston ", "name": "July 2018" },
    { "category": "one", "element_id": "august2018ONE-Boston-1", "location": "Boston ", "name": "August 2018" }
  ],
  "m": [
    { "category": "three", "element_id": "june2018THREE-Boston-1", "location": "Boston ", "name": "June 2018" },
    { "category": "two", "element_id": "july2018TWO-Boston-1", "location": "Boston ", "name": "July 2018" },
    { "category": "three", "element_id": "august2014THREE-Boston-1", "location": "Boston ", "name": "August 2014" }
  ]
}

test("MemberIcon", () => {
  let component = renderer.create(
    <MemberIcon member="d" summaryCounts={summaryCounts} reunionsForMember={reunionsForMember}></MemberIcon>
  )
  expect(component).toMatchSnapshot()
})

