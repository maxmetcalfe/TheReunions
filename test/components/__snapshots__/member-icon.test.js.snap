// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberIcon 1`] = `
<div
  className="member-icon"
  onClick={[Function]}
>
  D
</div>
`;
